lib/src/local_working_copy.rs

```diff
@@ -1218,6 +1212,44 @@ impl FileSnapshotter<'_> {
         Ok(())
     }
 
+    /// Emits file paths that don't exist in the `present_entries`.
+    fn emit_deleted_files(
+        &self,
+        dir: &RepoPath,
+        file_states: FileStates<'_>,
+        present_entries: &PresentDirEntries,
+    ) {
+        #[derive(Clone, Copy, Debug, Eq, PartialEq)]
+        enum Kind {
+            Dir,
+            File,
+        }
+
+        let file_state_chunks = file_states.iter().chunk_by(|(path, _state)| {
+            // Extract <name> from <dir>, <dir>/<name>, or <dir>/<name>/**.
+            // (file_states may contain <dir> file on file->dir transition.)
+            debug_assert!(path.starts_with(dir));
+            let len = dir.as_internal_file_string().len() + (!dir.is_root() as usize);
+            let tail = path.as_internal_file_string().get(len..).unwrap_or("");
```

Comment:
suggestion: Maybe a variable name like `suffix` or `relative_path` is more helpful than `tail` (if that's correct)?

Reply:
I avoided `suffix` because it sounded like a file suffix. It could be called a `relative_path`, but it's a bit odd that `path` is `RepoPath`, but `relative_path` is `str`.